core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -204,6 +209,10 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
   private var receivedTransactionalOffsetCommits = false
   private var receivedConsumerOffsetCommits = false
 
+  // When protocolType == `consumer`, a set of subscribed topics is maintained. The set is
+  // computed when a new generation is created or when the group is restored from the log.
```

Comment:
So this means that offset deletion does not take into account the subscriptions of an active rebalance. That's probably fine I guess since it is simpler and we are not modifying the commit state for any partitions which are actively being fetched.

A couple questions for completeness:

1. For the sake of argument, suppose we loosened the validation so that we allow explicit deletions (but not expirations) without respect to the group or subscription state. The behavior of the consumer would be a little unpredictable. It might just write over the deletion. It might also force an offset reset if an OffsetFetch followed the deletion. Because the behavior is undefined, we do not allow it, which seems fair. However, the downside is that it makes the API a little harder to use, so just wanted to check whether you think it is a worthwhile restriction.

2. If we know the subscription, then it wouldn't be much trouble to delete the offsets of all partitions that are not currently subscribed to. This could be supported in the API by using a null in the OffsetDelete request. I think it could be useful, but I'm inclined to leave it for future work. What do you think?

Reply:
Yes and No. I have distinguished two cases when I designed this:

1) When the group is empty and the initial rebalance takes place, subscriptions will be ignored until the rebalance is completed and a new generation is initialised. I thought it would be inappropriate to compute the set of subscribed topics before having the full view of the consumer group. 

2) When the group is not empty and a rebalance takes place, the current set of subscribed topics will remain until all rebalance is completed and a new generation is initialised. I think this is a good think to be conservative here until the group has a full view of all the members.

Another aspect is that the protocol is actually selected at this point as well. I think it is better to stay consistent and parse the consumer protocol once the protocol of the group has been selected. It is not a hard requirement though as we could parse any versions but it feels more consistent in my opinion.

To answer your questions:

1) I think allowing explicit deletions without respect to the group or subscription state would be really risky as the behaviour of the consumer would be unpredictable as you pointed out. I would say that it does not make much sense anyway to delete offsets of a topic while a running consumer group consumes this topic. It is good to have the restriction in place in my opinion.

2) Yeah. It would be possible to delete the offsets of all partitions *of the topics* that are not currently subscribed to. I wonder it it really useful though as they will be eventually cleaned up now. I don't see the benefit of having the possibility to basically force the immediate clean up. I would leave it for future work and re-iterate based on the feedback we get from the users.

Reply:
Thanks for the response. Makes sense and I agree.